# Multiple Sequence Alignment Project

This project aims to perform multiple sequence alignment for given k protein amino acid sequences.

## Project Files

- **Project.docx**: General documentation of the project.
- **Blosum62.txt**: File containing the BLOSUM62 matrix.
- **Program.py**: Source code of the project written in Python.
- **Input.txt**: Example input file.
- **Input2.txt**: Second example input file.
- **Input3.txt**: Third example input file.
- **Multiple-Alignment.txt**: Optimal multiple sequence alignment result for the first example input file.
- **Multiple-Alignment2.txt**: Optimal multiple sequence alignment result for the second example input file.
- **Multiple-Alignment3.txt**: Optimal multiple sequence alignment result for the third example input file.

## Project Description

The main goal of the project is to perform multiple sequence alignment for a certain number of protein amino acid sequences. This process is carried out using the given BLOSUM62 matrix and a gap penalty obtained from the user.

**Note:** `Multiple-Alignment.txt` files contain optimal multiple sequence alignment results.

## Running the Project

1. Run the `Program.py` file in a Python environment.
2. Enter the gap penalty.
3. The program will read input files and perform multiple sequence alignment.
4. You can examine the results of multiple sequence alignment.
5. Additionally, view the generated phylogenetic tree to analyze the evolutionary relationships between sequences.
